go/test/endtoend/cluster/cluster_process.go

```diff
@@ -0,0 +1,338 @@
+	log.Info(fmt.Sprintf("Starting vtgate on port %d", vtgateHTTPPort))
+	cluster.VtgateProcess = *VtgateProcessInstance(
+		vtgateHTTPPort,
+		vtgateGrpcPort,
+		cluster.VtgateMySQLPort,
+		cluster.Cell,
+		cluster.Cell,
+		cluster.Hostname,
+		"MASTER,REPLICA",
+		cluster.topoProcess.Port,
+		cluster.TmpDirectory,
+		cluster.VtGateExtraArgs)
+
+	log.Info(fmt.Sprintf("Vtgate started, connect to mysql using : mysql -h 127.0.0.1 -P %d", cluster.VtgateMySQLPort))
+	return cluster.VtgateProcess.Setup()
+}
+
+// NewCluster instantiates a new cluster
+func NewCluster(cell string, hostname string) *LocalProcessCluster {
+	cluster := &LocalProcessCluster{Cell: cell, Hostname: hostname}
+	cluster.OriginalVTDATAROOT = os.Getenv("VTDATAROOT")
+	cluster.CurrentVTDATAROOT = path.Join(os.Getenv("VTDATAROOT"), fmt.Sprintf("vtroot_%d", cluster.GetAndReservePort()))
+	_ = createDirectory(cluster.CurrentVTDATAROOT, 0700)
+	_ = os.Setenv("VTDATAROOT", cluster.CurrentVTDATAROOT)
+	return cluster
+}
+
+// ReStartVtgate starts vtgate with updated configs
+func (cluster *LocalProcessCluster) ReStartVtgate() (err error) {
+	err = cluster.VtgateProcess.TearDown()
+	if err != nil {
+		log.Error(err.Error())
+		return
+	}
+	err = cluster.StartVtgate()
+	if err != nil {
+		log.Error(err.Error())
+		return
+	}
+	return err
+}
+
+// Teardown brings down the cluster by invoking teardown for individual processes
+func (cluster *LocalProcessCluster) Teardown() (err error) {
+	if err = cluster.VtgateProcess.TearDown(); err != nil {
+		log.Error(err.Error())
+		return
+	}
+
+	for _, keyspace := range cluster.Keyspaces {
+		for _, shard := range keyspace.Shards {
+			for _, tablet := range shard.Vttablets {
+				if err = tablet.mysqlctlProcess.Stop(); err != nil {
+					log.Error(err.Error())
+					return
+				}
+
+				if err = tablet.vttabletProcess.TearDown(); err != nil {
+					log.Error(err.Error())
+					return
+				}
+			}
+		}
+	}
+
+	if err = cluster.vtctldProcess.TearDown(); err != nil {
+		log.Error(err.Error())
+		return
+	}
+
+	if err = cluster.topoProcess.TearDown(cluster.Cell, cluster.OriginalVTDATAROOT, cluster.CurrentVTDATAROOT, *keepData); err != nil {
+		log.Error(err.Error())
+		return
+	}
+	return err
+}
+
+// GetAndReservePort gives port for required process
+func (cluster *LocalProcessCluster) GetAndReservePort() int {
+	if cluster.nextPortForProcess == 0 {
+		cluster.nextPortForProcess = getRandomNumber(20000, 15000)
+	}
+	cluster.nextPortForProcess = cluster.nextPortForProcess + 1
+	return cluster.nextPortForProcess
+}
+
+// GetAndReserveTabletUID gives tablet uid
+func (cluster *LocalProcessCluster) GetAndReserveTabletUID() int {
+	if cluster.BaseTabletUID == 0 {
+		cluster.BaseTabletUID = getRandomNumber(10000, 0)
+	}
+	cluster.BaseTabletUID = cluster.BaseTabletUID + 1
+	return cluster.BaseTabletUID
+}
+
+func getRandomNumber(maxNumber int32, baseNumber int) int {
+	return int(rand.Int31n(maxNumber)) + baseNumber
```

Comment:
This is not as random as you think. At minimum, you have to call `rand.Seed`. Otherwise, all processes will generate the exact same sequence of random numbers.

I'm also worried about the possibility of collisions even after seeding. But I don't have an elegant answer right now.

Reply:
Currently all these type of tests run sequentially - [check here](https://github.com/planetscale/vitess/blob/cluster_for_test/tools/e2e_test_cluster.sh#L28). However, I noticed that for every run of test, the processes generated with different numbers for port even in travis and github workflow (just verified those in execution log) as well as in local environment (Whereas it should have generated same seq of random numbers for every run). Anyway I will try with `rand.Seed` to add some randomness.

Reply:
Added `rand.seed` in the recent commit.